### PR TITLE
ui: Change time selector to a dropdown

### PR DESCRIPTION
**Pull request by maxlang**
This matches the time selector to the current designs:
- time selection is now a dropdown
- the currently selected time range is now visible at all times
- you can shift the timerange backwards and forwards in time using
  the ui widget and the custom time range is shown

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/12360)

### CODE REVIEW

**Comment by maxlang** [comment]
Range > 48 hours
screen shot 2016-12-14 at 4 56 49 pm

Range < 48 hours on the same day
screen shot 2016-12-14 at 5 45 29 pm

Range < 48 hours today 
screen shot 2016-12-14 at 6 04 21 pm

Range < 48 hrs spanning multiple days
screen shot 2016-12-14 at 4 56 27 pm

Non-custom range
screen shot 2016-12-14 at 4 56 13 pm

**Comment by kuanluo** [comment]
LGTM!

**Comment by vivekmenezes** [comment]
LGTM

File: pkg/ui/app/components/dropdown.tsx [hunk 1]
```diff
@@ -20,8 +31,10 @@ export default class Selector extends React.Component<SelectorOwnProps, {}> {
   render() {
     let {selected, options, onChange} = this.props;
     return <div className="dropdown">
-      <span className="dropdown__title">{ this.props.title }:</span>
+      <span className="Select__side-arrow" dangerouslySetInnerHTML={trustIcon(leftArrow)} onClick={() => this.props.onArrowClick(ArrowDirection.LEFT)}></span>
```
Comment (mrtracy) [inline pkg/ui/app/components/dropdown.tsx]: This should be `dropdown__side-arrow`?  This is not inside of the `Select` element.

I'm not thrilled about the arrows being present on all dropdowns, but you put the relevant CSS in the shame file so I guess that's fine for now.  Maybe a TODO here?

File: pkg/ui/app/components/dropdown.tsx [hunk 2]
```diff
@@ -20,8 +31,10 @@ export default class Selector extends React.Component<SelectorOwnProps, {}> {
   render() {
```
Comment (mrtracy) [inline pkg/ui/app/components/dropdown.tsx]: This component should be called "Dropdown" everywhere.  We are mixing "Selector" and "Dropdown", and that's very confusing when combined with the third-party "Select" component we are using.

File: pkg/ui/app/containers/timescale.tsx [hunk 3]
```diff
@@ -29,46 +66,61 @@ interface TimeScaleSelectorProps {
...
+        selected.windowEnd.subtract(selected.windowSize);
+        break;
+      default:
+        console.error("Unknown direction: ", direction);
+    }
 
-  hide = (e: Event) => {
-    if (e.target !== this.timescaleBtn) {
-      this.setVisible(false);
+    if (!selected.windowEnd || selected.windowEnd > moment().subtract(selected.windowValid)) {
```
Comment (mrtracy) [inline pkg/ui/app/containers/timescale.tsx]: Please add a comment here about what this is doing (making the time window auto-update again).

File: pkg/ui/app/containers/timescale.tsx [hunk 4]
```diff
@@ -29,46 +66,61 @@ interface TimeScaleSelectorProps {
...
+    this.props.setTimeScale(selected);
   }
 
-  isSelected(scale: timewindow.TimeScale) {
-    return scale === this.props.currentScale;
+  getTimescaleOptions = () => {
+    let timescaleOptions = _.map(timewindow.availableTimeScales, (ts, k) => {
+      return { value: k, label: "Last " + k };
+    });
+    if (!_.has(timewindow.availableTimeScales, this.props.currentScale.key)) {
```
Comment (mrtracy) [inline pkg/ui/app/containers/timescale.tsx]: It took me a bit to figure out what this was doing with regards to the "Custom" selection value.  I think your solution is good, but please add a comment about handling the "Custom" group if present.

File: pkg/ui/styl/base/palette.styl [hunk 5]
```diff
@@ -88,4 +87,6 @@ $glow-blue = #51BBFD
 $lighter-blue = #BEBECD
 $light-gray = #B8BBBC
 $light-gray-2 = #DCDCDC
-$lighter-gray = #F2F2F2
\ No newline at end of file
+$lighter-gray = #F2F2F2
+$grey-blue = #C8CBD4
+$dark-grey-blue = #728099
```
Comment (mrtracy) [inline pkg/ui/styl/base/palette.styl]: add newline at end of file.

File: pkg/ui/app/components/dropdown.tsx [hunk 6]
```diff
@@ -1,27 +1,41 @@
...
+interface DropdownOwnProps {
   title: string;
   selected: string;
-  options: SelectorOption[];
-  onChange?: (selected: SelectorOption) => void; // Callback when the value changes.
+  options: DropdownOption[];
+  onChange?: (selected: DropdownOption) => void; // Callback when the value changes.
+  // If true, don't display the arrow next to the dropdown, display left and
+  // right arrows to either side instead.
+  sideArrows?: boolean;
```
Comment (mrtracy) [inline pkg/ui/app/components/dropdown.tsx]: You are not using this property.

File: pkg/ui/app/containers/timescale.tsx [hunk 7]
```diff
@@ -86,50 +144,42 @@ class TimeScaleSelector extends React.Component<TimeScaleSelectorProps, TimeScal
...
-                           onClick={(e) => { this.changeSettings(theseSettings); this.setVisible(false); } }>
-                           { key }
-            </button>;
-          })
-        }
-      </div>
+    let currentScale = this.props.currentScale;
+    let timeRange: React.ReactNode = "";
+
+    let className = "Select--side-arrows";
```
Comment (mrtracy) [inline pkg/ui/app/containers/timescale.tsx]: This is a really odd decision. "Select" is the name of the CSS class applied to the react-select component.  You've created html here that looks like:

```html
<div class="Select--side-arrows">
  <div class="dropdown">
     <div class="Select dropdown__select">
```

That's very difficult to comprehend.  

Why isn't side arrow visibility handled by the dropdown component?  You even added an "arrows" boolean property to the dropdown, but then didn't use it, instead electing to put the arrows on every single dropdown but keep them hidden, and then overriding that hidden behavior using this oddly-named CSS class in the time window component.

File: pkg/ui/styl/base/palette.styl (hunk 5 above)
Comment (maxlang) [inline pkg/ui/styl/base/palette.styl]: Done.

**Review by maxlang — COMMENTED** [review]
@mrtracy Arrow visibility is now handled by the dropdown component. As is whether the arrow is disabled.

@kuanluo The right arrow is now disabled instead of removed. I didn't know which colors to use so this is what it looks like currently.
screen shot 2016-12-18 at 10 28 51 pm

**Comment by mrtracy** [comment]
LGTM, thanks for the fixes

**Pull request merged**